File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> v;
        int o=n;
        int c=n;
        solve(o,c,"",v);
        return v;
        }
    
    void solve(int o,int c,string ans,vector<string>& v)
        {
        if(o==0 && c==0)
            { v.push_back(ans);
             return;
             }
        
        if(o>0)
            { solve(o-1,c,ans+'(',v);
             }
        if(c>o)
            { solve(o,c-1,ans+')',v);
             }
             
             
        
        
    
 
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.dart
class Solution {
  List<String> generateParenthesis(int n) {
  List<String> result = [];

  void generate(String p, int left, int right) {
    if (left == 0 && right == 0) {
      result.add(p);
      return;
    }

    if (left > 0) {
      generate(p + '(', left - 1, right);
    }

    if (right > left) {
      generate(p + ')', left, right - 1);
    }
  }

  generate('', n, n);
  return result;
}

}